File: intrinsic/assets/services/examples/calcserver/calc_server.cc
// Copyright 2023 Intrinsic Innovation LLC

#include "intrinsic/assets/services/examples/calcserver/calc_server.h"

#include <cstdint>
#include <memory>
#include <string>
#include <utility>

#include "absl/log/log.h"
#include "absl/status/status.h"
#include "absl/strings/str_cat.h"
#include "grpcpp/channel.h"
#include "grpcpp/client_context.h"
#include "grpcpp/server_context.h"
#include "grpcpp/support/status.h"
#include "intrinsic/assets/dependencies/utils.h"
#include "intrinsic/assets/proto/v1/resolved_dependency.pb.h"
#include "intrinsic/assets/services/examples/calcserver/calc_server.grpc.pb.h"
#include "intrinsic/assets/services/examples/calcserver/calc_server.pb.h"
#include "intrinsic/resources/proto/runtime_context.pb.h"
#include "intrinsic/util/status/status_conversion_grpc.h"
#include "intrinsic/util/status/status_macros.h"
#include "intrinsic/util/status/status_macros_grpc.h"

namespace intrinsic::services {

namespace {

char kCustomCalculationInterface[] =
    "grpc://intrinsic_proto.services.CustomCalculation";
char kConversionDatasetInterface[] =
    "data://intrinsic_proto.services.ConversionDataset";

// Retrieves the linear conversion parameters (factor and offset) for a given
// unit. Returns NotFoundError if the unit is not found (null), or
// UnimplementedError if the conversion type is not supported.
absl::StatusOr<std::pair<double, double>> GetLinearParams(
    const intrinsic_proto::services::Unit* unit, absl::string_view unit_name) {
  if (unit) {
    if (!unit->has_linear()) {
      return absl::UnimplementedError(
          absl::StrCat("unsupported conversion type for unit: ", unit_name));
    }
    return std::make_pair(unit->linear().factor(), unit->linear().offset());
  }
  return absl::NotFoundError(absl::StrCat("unit not found: ", unit_name));
}

}  // namespace

grpc::Status CalculatorServiceImpl::Calculate(
    grpc::ServerContext* context,
    const intrinsic_proto::services::CalculatorRequest* request,
    intrinsic_proto::services::CalculatorResponse* response) {
  int64_t result;

  int64_t a, b;
  if (config_.reverse_order()) {
    a = request->y();
    b = request->x();
  } else {
    a = request->x();
    b = request->y();
  }

  switch (request->operation()) {
    case intrinsic_proto::services::CALCULATOR_OPERATION_ADD:
      result = a + b;
      LOG(INFO) << a << " + " << b << " = " << result;
      break;
    case intrinsic_proto::services::CALCULATOR_OPERATION_MULTIPLY:
      result = a * b;
      LOG(INFO) << a << " * " << b << " = " << result;
      break;
    case intrinsic_proto::services::CALCULATOR_OPERATION_SUBTRACT:
      result = a - b;
      LOG(INFO) << a << " - " << b << " = " << result;
      break;
    case intrinsic_proto::services::CALCULATOR_OPERATION_DIVIDE:
      if (b == 0) {
        LOG(INFO) << "Cannot divide by 0 (" << a << " / " << b << ")";
        return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
                            "Cannot divide by 0");
      }
      result = a / b;
      LOG(INFO) << a << " / " << b << " = " << result;
      break;
    case intrinsic_proto::services::CALCULATOR_OPERATION_CUSTOM: {
      // Connect to the CustomCalculation service.
      ::grpc::ClientContext ctx;
      INTR_ASSIGN_OR_RETURN_GRPC(
          std::shared_ptr<grpc::Channel> channel,
          assets::dependencies::Connect(ctx, config_.custom_calculation(),
                                        kCustomCalculationInterface));
      auto stub =
          intrinsic_proto::services::CustomCalculation::NewStub(channel);

      // Create the request.
      intrinsic_proto::services::CustomCalculateRequest custom_request;
      custom_request.set_x(a);
      custom_request.set_y(b);

      // Call the CustomCalculation service.
      intrinsic_proto::services::CalculatorResponse custom_response;
      INTR_RETURN_IF_ERROR_GRPC(
          stub->Calculate(&ctx, custom_request, &custom_response));
      result = custom_response.result();
      LOG(INFO) << "Custom(" << a << ", " << b << ") = " << result;

      break;
    }
    default:
      return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
                          "Invalid operation");
  }
  response->set_result(result);
  return grpc::Status::OK;
}

absl::StatusOr<intrinsic_proto::services::ConversionDataset>
CalculatorServiceImpl::GetConversionDataset() {
  if (!config_.has_conversion_dataset()) {
    return absl::FailedPreconditionError("conversion dataset not configured");
  }

  INTR_ASSIGN_OR_RETURN(auto payload, fetcher_(config_.conversion_dataset(),
                                               kConversionDatasetInterface));

  intrinsic_proto::services::ConversionDataset dataset;
  if (!payload.UnpackTo(&dataset)) {
    return absl::InternalError("failed to unmarshal dataset");
  }

  return dataset;
}

grpc::Status CalculatorServiceImpl::Convert(
    grpc::ServerContext* context,
    const intrinsic_proto::services::ConvertRequest* request,
    intrinsic_proto::services::ConvertResponse* response) {
  if (request->category().empty()) {
    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
                        "category not specified");
  }
  if (request->from_unit().empty()) {
    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
                        "from unit not specified");
  }
  if (request->to_unit().empty()) {
    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
                        "to unit not specified");
  }

  INTR_ASSIGN_OR_RETURN_GRPC(
      const intrinsic_proto::services::ConversionDataset& dataset,
      GetConversionDataset());

  // Find the requested category within the dataset.
  const intrinsic_proto::services::UnitCategory* target_category_ptr = nullptr;
  for (const intrinsic_proto::services::UnitCategory& category :
       dataset.categories()) {
    if (category.name() == request->category()) {
      target_category_ptr = &category;
      break;
    }
  }
  if (!target_category_ptr) {
    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
                        "category not found: " + request->category());
  }

  // Check that the base unit is not redefined in the units list.
  for (const intrinsic_proto::services::Unit& unit :
       target_category_ptr->units()) {
    if (unit.name() == target_category_ptr->base_unit()) {
      return grpc::Status(grpc::StatusCode::FAILED_PRECONDITION,
                          "base unit redefined in units list: " + unit.name());
    }
  }

  // Create a mutable copy to add the base unit manually.
  intrinsic_proto::services::UnitCategory target_category =
      *target_category_ptr;
  intrinsic_proto::services::Unit* new_unit = target_category.add_units();
  new_unit->set_name(target_category.base_unit());
  new_unit->mutable_linear()->set_factor(1.0);
  new_unit->mutable_linear()->set_offset(0.0);

  const intrinsic_proto::services::Unit* from_proto_unit = nullptr;
  const intrinsic_proto::services::Unit* to_proto_unit = nullptr;

  for (const intrinsic_proto::services::Unit& unit : target_category.units()) {
    if (unit.name() == request->from_unit()) {
      from_proto_unit = &unit;
    }
    if (unit.name() == request->to_unit()) {
      to_proto_unit = &unit;
    }
  }

  INTR_ASSIGN_OR_RETURN_GRPC(
      (std::pair<double, double> from_params),
      GetLinearParams(from_proto_unit, request->from_unit()));
  INTR_ASSIGN_OR_RETURN_GRPC(
      (std::pair<double, double> to_params),
      GetLinearParams(to_proto_unit, request->to_unit()));

  double from_factor = from_params.first;
  double from_offset = from_params.second;
  double to_factor = to_params.first;
  double to_offset = to_params.second;
  if (config_.reverse_order()) {
    std::swap(from_factor, to_factor);
    std::swap(from_offset, to_offset);
  }

  if (to_factor == 0) {
    return grpc::Status(
        grpc::StatusCode::FAILED_PRECONDITION,
        "invalid factor 0 in dataset for unit: " + request->to_unit());
  }

  double value = request->value();
  double base_val = (value * from_factor) + from_offset;
  double result = (base_val - to_offset) / to_factor;

  LOG(INFO) << value << " " << request->from_unit() << " = " << result << " "
            << request->to_unit();
  response->set_result(result);
  return grpc::Status::OK;
}

}  // namespace intrinsic::services

File: intrinsic/assets/services/examples/calcserver/calc_server.py
# Copyright 2023 Intrinsic Innovation LLC

"""Implementation of a basic calculator service."""

from __future__ import annotations

from typing import Any

from absl import logging
import grpc

from intrinsic.assets.dependencies import utils
from intrinsic.assets.services.examples.calcserver import calc_server_pb2
from intrinsic.assets.services.examples.calcserver import calc_server_pb2_grpc

_CONVERSION_DATASET_INTERFACE = (
    'data://intrinsic_proto.services.ConversionDataset'
)


class CalculatorServiceServicer(calc_server_pb2_grpc.CalculatorServicer):
  """Performs basic calculator operations."""

  def __init__(self, config: calc_server_pb2.CalculatorConfig):
    """Initializes the object.

    Args:
      config: The calculator configuration.
    """
    self._config = config

  def _get_conversion_dataset(
      self,
      context: grpc.ServicerContext,
  ) -> calc_server_pb2.ConversionDataset:
    """Fetches the conversion dataset from the config."""
    if not self._config.HasField('conversion_dataset'):
      context.abort(
          grpc.StatusCode.FAILED_PRECONDITION,
          'Conversion dataset dependency not configured',
      )

    dataset = calc_server_pb2.ConversionDataset()
    try:
      any_payload = utils.get_data_payload(
          self._config.conversion_dataset, _CONVERSION_DATASET_INTERFACE
      )
      any_payload.Unpack(dataset)
    except grpc.RpcError as e:
      context.abort(e.code(), e.details())
    except Exception as e:
      context.abort(grpc.StatusCode.INTERNAL, f'failed to get dataset: {e}')

    return dataset

  def _get_linear_params(
      self,
      unit: calc_server_pb2.Unit | None,
      unit_name: str,
      context: grpc.ServicerContext,
  ) -> tuple[float, float]:
    """Retrieves the linear conversion parameters with respect to the base unit."""
    if unit:
      if not unit.HasField('linear'):
        context.abort(
            grpc.StatusCode.UNIMPLEMENTED,
            f'unsupported conversion type for unit: {unit_name}',
        )
      return unit.linear.factor, unit.linear.offset

    context.abort(grpc.StatusCode.NOT_FOUND, f'unit not found: {unit_name}')

  def Calculate(
      self,
      request: calc_server_pb2.CalculatorRequest,
      context: grpc.ServicerContext,
  ) -> calc_server_pb2.CalculatorResponse:
    result = 0

    if self._config.reverse_order:
      a = request.y
      b = request.x
    else:
      a = request.x
      b = request.y

    if request.operation == calc_server_pb2.CALCULATOR_OPERATION_ADD:
      result = a + b
      logging.info('%d + %d = %d', a, b, result)
    elif request.operation == calc_server_pb2.CALCULATOR_OPERATION_MULTIPLY:
      result = a * b
      logging.info('%d * %d = %d', a, b, result)
    elif request.operation == calc_server_pb2.CALCULATOR_OPERATION_SUBTRACT:
      result = a - b
      logging.info('%d - %d = %d', a, b, result)
    elif request.operation == calc_server_pb2.CALCULATOR_OPERATION_DIVIDE:
      if b == 0:
        logging.info('Cannot divide by 0 (%d / %d)', a, b)
        context.abort(grpc.StatusCode.INVALID_ARGUMENT, 'Cannot divide by 0')
      result = a // b
      logging.info('%d / %d = %d', a, b, result)
    else:
      context.abort(
          grpc.StatusCode.UNIMPLEMENTED,
          f'Unsupported operation: {request.operation}',
      )

    return calc_server_pb2.CalculatorResponse(result=result)

  def Convert(
      self,
      request: calc_server_pb2.ConvertRequest,
      context: grpc.ServicerContext,
  ) -> calc_server_pb2.ConvertResponse:
    # Validation
    if not request.category:
      context.abort(grpc.StatusCode.INVALID_ARGUMENT, 'category not specified')
    if not request.from_unit:
      context.abort(
          grpc.StatusCode.INVALID_ARGUMENT,
          'from unit not specified',
      )
    if not request.to_unit:
      context.abort(grpc.StatusCode.INVALID_ARGUMENT, 'to unit not specified')

    dataset = self._get_conversion_dataset(context)

    target_category = None
    for category in dataset.categories:
      if category.name == request.category:
        target_category = category
        break

    if not target_category:
      context.abort(
          grpc.StatusCode.INVALID_ARGUMENT,
          f'category not found: {request.category}',
      )

    # Check that the base unit is not redefined in the units list.
    for unit in target_category.units:
      if unit.name == target_category.base_unit:
        context.abort(
            grpc.StatusCode.FAILED_PRECONDITION,
            f'base unit {unit.name} redefined in units list',
        )

    # Add the base unit manually to the units list to simplify lookup.
    base_unit_proto = calc_server_pb2.Unit(
        name=target_category.base_unit,
        linear=calc_server_pb2.LinearConversion(factor=1.0, offset=0.0),
    )
    target_category.units.append(base_unit_proto)

    from_unit = None
    to_unit = None
    for unit in target_category.units:
      if unit.name == request.from_unit:
        from_unit = unit
      if unit.name == request.to_unit:
        to_unit = unit

    from_factor, from_offset = self._get_linear_params(
        from_unit, request.from_unit, context
    )
    to_factor, to_offset = self._get_linear_params(
        to_unit, request.to_unit, context
    )

    if self._config.reverse_order:
      from_factor, to_factor = to_factor, from_factor
      from_offset, to_offset = to_offset, from_offset

    if to_factor == 0:
      context.abort(
          grpc.StatusCode.FAILED_PRECONDITION,
          f'invalid factor 0 for unit: {request.to_unit}',
      )

    value = request.value
    base_val = (value * from_factor) + from_offset
    result = (base_val - to_offset) / to_factor

    logging.info(
        '%f %s = %f %s', value, request.from_unit, result, request.to_unit
    )
    return calc_server_pb2.ConvertResponse(result=result)
